#!/bin/bash	
#PBS -N bwa-PA2013-001
#PBS -l nodes=1:ppn=16
#PBS -l vmem=100gb
#PBS -l walltime=16:00:00
#PBS -M ouqd@hotmail.com
#PBS -m abe
#PBS -j oe

# This pipeline pbs is produced by the perl script:
# perl MGMP.pl bwa PA42.3.0 /N/dc2/scratch/xw63/PA2013
# Date and time: Wed Jul 11 12:36:13 2018

set +x
module load samtools
module load java
ulimit -s
set -x
cd /N/u/xw63/Carbonate/daphnia/genome_index
mkdir /N/dc2/scratch/xw63/PA2013/Bwa
mkdir /N/dc2/scratch/xw63/PA2013/tmp

set +x
echo ===============================================================
echo 0. making index files, which should be done before submitting pbs
echo ===============================================================
echo samtools faidx PA42.3.0.fasta
echo bwa index PA42.3.0.fasta 
echo hisat2-build PA42.3.0.fasta PA42.3.0 
echo /N/soft/rhel6/novoalign/novocraft/novoindex PA42.3.0.ndx PA42.3.0.fasta 
echo rm PA42.3.0.dict
echo java -XX:ParallelGCThreads=4 -Xmx32g -Xms32g -Djavaio.tmpdir=/N/dc2/scratch/xw63/PA2013/tmp -jar /N/soft/rhel6/picard/2.8.1/picard.jar  CreateSequenceDictionary R=PA42.3.0.fasta O=PA42.3.0.dict
echo These commands should be executed before submitting this pbs.
echo DO NOT excute these commands repeatedly in the pbs jobs, as it will cause problems when one job is using the index while another job is re-creating the index.
echo ===============================================================
echo 1. After preparing the FASTA file of adapter sequences, trim adapter sequences from sequence reads.
echo ===============================================================

set -x
module load java

time java -XX:ParallelGCThreads=4 -Xmx32g -Xms32g -Djavaio.tmpdir=/N/dc2/scratch/xw63/PA2013/tmp -jar ~/Trimmomatic-0.36/trimmomatic-0.36.jar PE /N/dc2/scratch/xw63/PA2013/fastq/PA2013-001-R1.fastq /N/dc2/scratch/xw63/PA2013/fastq/PA2013-001-R2.fastq /N/dc2/scratch/xw63/PA2013/fastq/PA2013-001-R1-paired.fq /N/dc2/scratch/xw63/PA2013/fastq/PA2013-001-R1-unpaired.fq /N/dc2/scratch/xw63/PA2013/fastq/PA2013-001-R2-paired.fq /N/dc2/scratch/xw63/PA2013/fastq/PA2013-001-R2-unpaired.fq HEADCROP:3 ILLUMINACLIP:/N/u/xw63/Carbonate/daphnia/Adapters/PA2013-001_Adapters.fa:2:30:10:2 SLIDINGWINDOW:4:15 MINLEN:30
set +x
echo ===============================================================
echo 3. Mapping reads to the reference sequence and output bam file.
echo ===============================================================

set -x

	
time ~/bwa-0.7.17/bwa mem -t 16 -M -k 30 PA42.3.0.fasta /N/dc2/scratch/xw63/PA2013/fastq/PA2013-001-R1-paired.fq /N/dc2/scratch/xw63/PA2013/fastq/PA2013-001-R2-paired.fq > /N/dc2/scratch/xw63/PA2013/Bwa/PA2013-001-paired.sam &
	
time ~/bwa-0.7.17/bwa mem -t 16 -M -k 30 PA42.3.0.fasta /N/dc2/scratch/xw63/PA2013/fastq/PA2013-001-R1-unpaired.fq > /N/dc2/scratch/xw63/PA2013/Bwa/PA2013-001-R1-unpaired.sam & 
	
time ~/bwa-0.7.17/bwa mem -t 16 -M -k 30 PA42.3.0.fasta /N/dc2/scratch/xw63/PA2013/fastq/PA2013-001-R2-unpaired.fq > /N/dc2/scratch/xw63/PA2013/Bwa/PA2013-001-R2-unpaired.sam &

wait 

set +x
echo ===============================================================
echo 4. Combine the SAM files using Picard.
echo ===============================================================

set -x
module load java

time java -XX:ParallelGCThreads=4 -Xmx32g -Xms32g -Djavaio.tmpdir=/N/dc2/scratch/xw63/PA2013/tmp -jar /N/soft/rhel6/picard/2.8.1/picard.jar MergeSamFiles I=/N/dc2/scratch/xw63/PA2013/Bwa/PA2013-001-paired.sam I=/N/dc2/scratch/xw63/PA2013/Bwa/PA2013-001-R1-unpaired.sam I=/N/dc2/scratch/xw63/PA2013/Bwa/PA2013-001-R2-unpaired.sam O=/N/dc2/scratch/xw63/PA2013/Bwa/PA2013-001.sam

set +x
echo ===============================================================
echo 5. Convert the SAM file to the BAM file.
echo ===============================================================

set -x
	
time /N/soft/rhel6/samtools/1.3.1/bin/samtools view -bS /N/dc2/scratch/xw63/PA2013/Bwa/PA2013-001.sam > /N/dc2/scratch/xw63/PA2013/Bwa/PA2013-001.bam

set +x
echo ===============================================================
echo 5-1. remove unwanted reads and multiple mapped reads.
echo ===============================================================
module load java
set -x

time /N/soft/rhel6/samtools/1.3.1/bin/samtools view -q 20 -f 3 -F 3844 -b /N/dc2/scratch/xw63/PA2013/Bwa/PA2013-001.bam > /N/dc2/scratch/xw63/PA2013/Bwa/PA2013-001-qFf.bam

set +x

 echo -f 3: remains only
 echo read paired: 0x1
 echo read mapped in proper pair: 0x2
 echo -F 3844: removesï¼š
 echo read unmapped: 0x4
 echo not primary alignment: 0x100
 echo read fails platform/vendor quality checks: 0x200
 echo read is PCR or optical duplicate: 0x400
 echo supplementary alignment: 0x800

set +x
echo ===============================================================
echo 6. Sort the BAM file using Picard.
echo ===============================================================

set -x
	
time java -XX:ParallelGCThreads=4 -Xmx32g -Xms32g -Djavaio.tmpdir=/N/dc2/scratch/xw63/PA2013/tmp -jar /N/soft/rhel6/picard/2.8.1/picard.jar SortSam INPUT=/N/dc2/scratch/xw63/PA2013/Bwa/PA2013-001-qFf.bam OUTPUT=/N/dc2/scratch/xw63/PA2013/Bwa/PA2013-001-qFf-Sorted.bam SORT_ORDER=coordinate
set +x

echo ===============================================================
echo 7. Add read groups to the sorted BAM file.
echo ===============================================================
module load java
set -x

time java -XX:ParallelGCThreads=4 -Xmx32g -Xms32g -Djavaio.tmpdir=/N/dc2/scratch/xw63/PA2013/tmp -jar /N/soft/rhel6/picard/2.8.1/picard.jar AddOrReplaceReadGroups INPUT=/N/dc2/scratch/xw63/PA2013/Bwa/PA2013-001-qFf-Sorted.bam OUTPUT=/N/dc2/scratch/xw63/PA2013/Bwa/PA2013-001-qFf-RG_Sorted.bam RGID=Daphnia RGLB=bar RGPL=illumina RGSM=/N/dc2/scratch/xw63/PA2013/PA2013-001 RGPU=6
set +x

echo ===============================================================
echo 8. Mark duplicate reads .
echo ===============================================================

set -x
module load java
time java -XX:ParallelGCThreads=4 -Xmx32g -Xms32g -Djavaio.tmpdir=/N/dc2/scratch/xw63/PA2013/tmp -jar /N/soft/rhel6/picard/2.8.1/picard.jar MarkDuplicates INPUT=/N/dc2/scratch/xw63/PA2013/Bwa/PA2013-001-qFf-RG_Sorted.bam OUTPUT=/N/dc2/scratch/xw63/PA2013/Bwa/PA2013-001-qFf-RG_Sorted_dedup.bam METRICS_FILE=/N/dc2/scratch/xw63/PA2013/Bwa/PA2013-001-qFf-metrics.txt

set +x  
echo ===============================================================
echo 9. Index the BAM file using Picard.
echo ===============================================================

module load java
set -x
time java -XX:ParallelGCThreads=4 -Xmx32g -Xms32g -Djavaio.tmpdir=/N/dc2/scratch/xw63/PA2013/tmp -jar /N/soft/rhel6/picard/2.8.1/picard.jar BuildBamIndex INPUT=/N/dc2/scratch/xw63/PA2013/Bwa/PA2013-001-qFf-RG_Sorted_dedup.bam

set +x
echo ===============================================================
echo 10. Define intervals to target for the local realignment.
echo ===============================================================

module load java
set -x
time java -XX:ParallelGCThreads=4 -Xmx32g -Xms32g -Djavaio.tmpdir=/N/dc2/scratch/xw63/PA2013/tmp -jar /N/soft/rhel6/gatk/3.4-0/GenomeAnalysisTK.jar -T RealignerTargetCreator -R PA42.3.0.fasta -I /N/dc2/scratch/xw63/PA2013/Bwa/PA2013-001-qFf-RG_Sorted_dedup.bam -o /N/dc2/scratch/xw63/PA2013/Bwa/PA2013-001.intervals

set +x
echo ===============================================================
echo 11. Locally realign reads around indels.
echo ===============================================================

module load java
set -x

time java -XX:ParallelGCThreads=4 -Xmx32g -Xms32g -Djavaio.tmpdir=/N/dc2/scratch/xw63/PA2013/tmp -jar /N/soft/rhel6/gatk/3.4-0/GenomeAnalysisTK.jar -T IndelRealigner -R PA42.3.0.fasta -I /N/dc2/scratch/xw63/PA2013/Bwa/PA2013-001-qFf-RG_Sorted_dedup.bam -targetIntervals /N/dc2/scratch/xw63/PA2013/Bwa/PA2013-001.intervals -o /N/dc2/scratch/xw63/PA2013/Bwa/PA2013-001-qFf-RG_Sorted_dedup_realigned.bam

set +x
echo ===============================================================
echo 12. Clip overlapping read pairs.
echo ===============================================================

set -x

time /N/soft/rhel6/bamUtil/1.0.13/bam clipOverlap --in /N/dc2/scratch/xw63/PA2013/Bwa/PA2013-001-qFf-RG_Sorted_dedup_realigned.bam --out /N/dc2/scratch/xw63/PA2013/Bwa/PA2013-001-qFf-RG_Sorted_dedup_realigned_Clipped.bam
set +x
echo ===============================================================
echo 13. Index the clipped BAM file using Samtools
echo ===============================================================

set -x

time /N/soft/rhel6/samtools/1.3.1/bin/samtools index /N/dc2/scratch/xw63/PA2013/Bwa/PA2013-001-qFf-RG_Sorted_dedup_realigned_Clipped.bam
set +x
echo ===============================================================
echo 14. Make the mpileup file from the BAM file.
echo ===============================================================

set -x
	
time /N/soft/rhel6/samtools/1.3.1/bin/samtools mpileup -f PA42.3.0.fasta /N/dc2/scratch/xw63/PA2013/Bwa/PA2013-001-qFf-RG_Sorted_dedup_realigned_Clipped.bam > /N/dc2/scratch/xw63/PA2013/Bwa/PA2013-001.mpileup

set +x
echo ===============================================================
echo =============Task completed.===================
echo ===============================================================
	
