#!/bin/bash 
#PBS -N mapgd-continue
#PBS -k o
#PBS -l nodes=1:ppn=16,walltime=96:00:00
#PBS -l vmem=100gb
#PBS -M ouqd@hotmail.com
#PBS -m abe
#PBS -j oe

# Updated on 05/28/2018

#trap 'echo "# $BASH_COMMAND"' DEBUG
set +x 
module rm gcc
module load gcc
module load gsl  

export DATA_DIR=/N/dc2/scratch/xw63/PA2013/Bwa
export HeaderFile=$DATA_DIR/PA42.header

module load samtools
set -x
cd $DATA_DIR
set +x
echo ===============================================================
echo 0. Make a header file
echo ===============================================================
set -x
echo samtools view -H $DATA_DIR/PA2013-001-RG_Sorted_dedup_realigned_Clipped.bam \> $HeaderFile
set +x
echo ===============================================================
echo 1. Make a pro file of nucleotide-read quartets -counts of A, C, G, and T, from the mpileup files of the clones.
echo ===============================================================
set -x
echo mapgd proview -i *.mpileup -H $HeaderFile \> PA2013.pro.txt
set +x

echo ===============================================================
echo 2. Exclude mtDNA data from the pro file.
echo ===============================================================
set -x

time grep -v '^PA42_mt_genome' PA2013.pro.txt > Nuc_PA2013.pro.txt
set +x

echo ===============================================================
echo 3. Run the allele command to estimate allele and genotype frequencies from the pro file.
echo ===============================================================
set -x
time mapgd allele -i Nuc_PA2013.pro.txt -o PA2013.map -p PA2013.clean
set +x

echo ===============================================================
echo 4. Run the filter command to filter the map file of ML estimates of the parameters.
echo ===============================================================
set -x
time mapgd filter -i PA2013.map.map -p 20 -q 0.05 -Q 0.45 -c 800 -C 2400 -o filtered_PA2013.map
set +x
echo -p: minimum value of the likelihood-ratio test statistic for polymorphism 
echo -q: minimum minor-allele frequency estimate
echo -Q: maximum minor-allele frequency estimate
echo -c: minimum population coverage
echo -C: maximum population coverage
echo ===============================================================
echo 5. Run the genotype command to generate a file of genotype likelihoods
echo ===============================================================
set -x
time mapgd genotype -p PA2013.clean.pro -m filtered_PA2013.map.map > PA2013.genotype
set +x

echo ===============================================================
echo Remove the unnecessary header and footer
echo ===============================================================
set -x
time awk '{if ($3 != "MN_FREQ" && $3 >= 0.0 && $3 <= 1.0) print}' PA2013.genotype > F_PA2013.genotype
set +x
echo ===============================================================
echo Randomly pick a specified number - 200000 of SNPs from the file of genotype likelihoods 
echo ===============================================================
set -x
time /N/dc2/projects/daphpops/Software/MAPGD-0.4.26/extras/sub_sample.py F_PA2013.genotype -N 200000 > 200K_F_PA2013.genotype
set +x
echo ===============================================================
echo Extract the header from the file of genotype likelihoods
echo ===============================================================

set -x

time head -n -1 PA2013.genotype | awk '{if ($3 == NULL || $1 ~ /^@/) print}' > header_PA2013.genotype
set +x
echo ===============================================================
echo Extract the footer from the file of genotype likelihoods
echo ===============================================================

set -x
time tail -n 1 PA2013.genotype > footer_PA2013.genotype
set +x
echo ===============================================================
echo Add the header and footer to the sub-sample of the file of genotype likelihoods
echo ===============================================================

set -x
time cat header_PA2013.genotype 200K_F_PA2013.genotype footer_PA2013.genotype > wh_wf_200K_F_PA2013.genotype
set +x
echo ===============================================================
echo 7. Run the relatedness command
echo ===============================================================

set -x
time mapgd relatedness -i wh_wf_200K_F_PA2013.genotype -o 200K_PA2013_rel.out

set +x

